framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,597 @@
+
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.tools.javac.code.Attribute.Array;
+import com.sun.tools.javac.code.Symbol.ClassSymbol;
+import com.sun.tools.javac.code.Symbol.TypeSymbol;
+import com.sun.tools.javac.code.Symbol.VarSymbol;
+import com.sun.tools.javac.code.Type;
+import com.sun.tools.javac.code.TypeAnnotationPosition;
+import com.sun.tools.javac.code.Type.ArrayType;
+import com.sun.tools.javac.code.Type.ClassType;
+
+/**
+ * SignatureInferenceScenes represents a set of annotations that are inferred
+ * in a program. It also writes those annotations into .jaif files.
+ * This class infers annotations for fields, method return types, and method
+ * parameters.
+ * The set of annotations inferred for a certain class is stored in an
+ * {@link annotations.el.AScene}, which will later be written into a .jaif file.
+ * For example, a class field of a class whose fully-qualified name is
+ * my.package.MyClass will have its inferred type stored in a Scene, and later
+ * written into a file named my.package.MyClass.jaif.
+ *
+ * Calling an update* method
+ * ({@link updateInferredFieldType}, {@link updateInferredMethodParametersTypes}, or
+ * {@link updateInferredMethodReturnType}) 
+ * reads the currently-stored type for an element in a Scene, if any,
+ * and replaces it by the LUB of it and the update method's argument.
+ *
+ * This class does not store annotations for an element if the element has
+ * explicit annotations - when calling an update* method passing as argument an
+ * explicitly annotated field, method return, or method parameter, it's inferred
+ * type is not stored in a Scene.
+ * TODO: We could add an option to update the type of explicitly annotated
+ * elements, but this currently is not recommended since the
+ * insert-annotations-to-source tool, which adds annotations from .jaif files
+ * into source code, adds annotations on top of existing
+ * annotations. See https://github.com/typetools/annotation-tools/issues/105
+ *
+ * @author pbsf
+ */
+public class SignatureInferenceScenes {
+
+    /**
+     * Path to where .jaif files will be written to and read from.
+     * This path is relative to where the CF's javac command is executed.
+     */
+    public final static String JAIF_FILES_PATH = "build/jaif-files/";
+
+    /** Maps file paths (Strings) to Scenes. */
+    private static Map<String, AScene> scenes = new HashMap<String, AScene>();
+
+    /** Set containing Scenes that were modified since the last time all
+     * Scenes were written into .jaif files. The String argument of this set
+     * is a path to the .jaif file of the corresponding Scene on the set. It
+     * is obtained by passing a class name as argument to the
+     * {@link getJaifPath} method.
+     *
+     * Modifying a Scene means adding (or changing) a type annotation for a
+     * field, method return type or method parameter type in the Scene.
```

Comment:
Trivial comment:  I prefer use of a serial comma before "or".  See https://homes.cs.washington.edu/~mernst/advice/write-technical-paper.html#serial-comma